storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoArray.java

```diff
@@ -605,8 +618,30 @@ public boolean nodeIsViableForHead(ProtoNode node) {
     if (node.isInvalid()) {
       return false;
     }
-    return doesCheckpointMatch(node.getJustifiedCheckpoint(), justifiedCheckpoint)
-        && doesCheckpointMatch(node.getFinalizedCheckpoint(), finalizedCheckpoint);
+
+    if (progressiveBalancesMode.isFull()) {
+      final boolean correctJustified;
+      final boolean correctFinalized;
+      final boolean isPreviousEpochJustified =
+          justifiedCheckpoint.getEpoch().plus(1).isGreaterThanOrEqualTo(currentEpoch);
+      if (isPreviousEpochJustified) {
+        correctJustified =
+            node.getJustifiedCheckpoint()
+                .getEpoch()
+                .isGreaterThanOrEqualTo(justifiedCheckpoint.getEpoch());
+        correctFinalized = correctJustified;
+      } else {
+        correctJustified =
+            node.getJustifiedCheckpoint()
+                .getEpoch()
+                .isGreaterThanOrEqualTo(justifiedCheckpoint.getEpoch());
```

Comment:
looks like this code is common, could be done outside the if?

Reply:
Yeah can do.